Added the FAQ section (#968)
thefrontside/effection: www/docs/faq.mdx

by blurring the line about where those guarantees apply. Right now, the line is very clear: anything on the 
right side of `yield*` is guaranteed to exit fully and never outlive the operation where the `yield*` was 
called. We can't make these guarantees with Promises for the reasons described in [The Await Event Horizon](https://frontside.com/blog/2023-12-11-await-event-horizon/)
blog post.

## Can I make Promise compatible with `yield*`?

Yes. You can `yield*` to a `Promise` by monkey patching the `Promise` prototype. "Monkey patching" in JavaScript refers to the practice 
of modifying or extending existing objects or functions at runtime, typically without changing their 
original source code. It's considered risky because it breaks the expectations that developers maintaining
and running the code have of the JavaScript runtime. There are circumstances where the convenience of `yield*` 
to promise outweighs the risks. We'll let you make the decision where those circumstances are for you. 

You can use the following code to monkeypatch Promise if you choose to do so. You can put this code into a module
and import it when you want this effect.

```typescript

> import { call, type Effect } from "effection";

Reviewer: ```suggestion
import { call, type Operation } from "effection";
```

Reviewer: We should probably test these....

Author: I'm going to move these to contrib

> declare global {
>  interface Promise<T> extends Operation<T> {}
> }
> 
> Object.defineProperty(Promise.prototype, Symbol.iterator, {
>   value(): {
>     return call(() => this)[Symbol.iterator]();
>   }
> });

Reviewer: ```suggestion
declare global {
  interface Promise<T> extends Operation<T> {}
}

Object.defineProperty(Promise.prototype, Symbol.iterator, {
  value() {
    return call(() => this)[Symbol.iterator]();
  },
});
```

Oops, actually tried running it, and this is what actually works

This page covers questions users of Effection ask as they go through the journey
of learning Effection.

## Why can't I `yield*` to a promise?

In TypeScript, if you tried to `yield* Promise.resolve(42)`, TypeScript compiler will
throw an error `Type 'Promise<number>' must have a '[Symbol.iterator]()' method that returns an iterator.`.

> In JavaScript, it will just not work. This happens because neither TypeScript nor JavaScript runtimes know how to 

Reviewer: ```suggestion
In JavaScript, it will just not work. This happens because neither the TypeScript nor the JavaScript runtimes know how to 
```

convert a Promise into an iterator, which is required for yield* to work.

> We opted to choose not to monkey patch the Promise prototype by adding the integration needed to make `yield*` 
> promise work, because this would compromise the guarantees that are described in [Thinking in Effection](./thinking-in-effection)
> by bluring the line about where the guarantees apply. Right now, the line is very clear, anything on the 
> right side of `yield*` is guaranteed to exit fully and never outlive the operation where the `yield*` was 
> called. We can't make these guarantees with Promises for the reasons described in [The Await Event Horizon](https://frontside.com/blog/2023-12-11-await-event-horizon/)
> blog post.

Reviewer: ```suggestion
While adding the integration to make `yield*` work with a promise is possible by monkey patching the `Promise` prototype, we opted not to do so. Not only is monkey patching a global object from a library considered bad form, but it would also compromise the guarantees that are described in [Thinking in Effection](./thinking-in-effection)
by blurring the line about where those guarantees apply. Right now, the line is very clear: anything on the 
right side of `yield*` is guaranteed to exit fully and never outlive the operation where the `yield*` was 
called. We can't make these guarantees with Promises for the reasons described in [The Await Event Horizon](https://frontside.com/blog/2023-12-11-await-event-horizon/)
blog post.
```

## Can I make Promise compatible with `yield*`?

> Yes, by monkeypatching the Promise prototype. Monkeypatching in JavaScript refers to the practice 
> of modifying or extending existing objects or functions at runtime, typically without changing their 
> original source code. It's considered risky because it breaks the expectations that developers maintaining
> and running the code have of the JavaScript runtime. There are circumstances where the convenience of `yield*` 
> to promise outweighs the risks. We'll let you make the decision where those circumstances are for you. 
> 
> You can use the following code to monkeypatch Promise if you choose to do so. You can put this code into a module
> and import it when you want this effect.

Reviewer: ```suggestion
Yes. You can `yield*` to a `Promise` by monkey patching the `Promise` prototype. "Monkey patching" in JavaScript refers to the practice 
of modifying or extending existing objects or functions at runtime, typically without changing their 
original source code. It's considered risky because it breaks the expectations that developers maintaining
and running the code have of the JavaScript runtime. There are circumstances where the convenience of `yield*` 
to promise outweighs the risks. We'll let you make the decision where those circumstances are for you. 

You can use the following code to monkeypatch Promise if you choose to do so. You can put this code into a module
and import it when you want this effect.
```

```typescript
import { call, type Effect } from "effection";

declare global {

>   interface Promise<T> {
>     [Symbol.iterator](): Iterator<Effect<unknown>, T, unknown>;
>   }

Reviewer: ```suggestion
 interface Promise<T> extends Operation<T> {}
```

`Effect<T>` is an internal  interface that should not be referenced directly. The most important thing to declare is that "promise is an operation". This will work with both v3 and v4

}

> Promise.prototype[Symbol.iterator] = function*<T>() {
>   return yield* call(() => this);
> }

Reviewer: There are a couple of differences I'd suggest: 

1. Don't put an enumerable property onto the prototype.
2. Remove a layer of delegation, so that you don't create an intermediate generator whose only purpose is to yield* to `call()`.
```suggestion
Object.defineProperty(Promise.prototype, Symbol.iterator, {
  value(): {
    return call(() => this)[Symbol.iterator]();
  }
});
```
